tests/test_extract.py

```diff
@@ -0,0 +1,145 @@
+        """
+        Test `zstash extract` with `--keep` applied on `create`, `update`, or `extract`.
+        """
+        self.hpss_path = hpss_path
+        use_hpss = self.setupDirs(test_name)
+        self.create(use_hpss, zstash_dir, keep=(keep=='create'))
+        self.add_files(use_hpss, zstash_dir, keep=(keep=='update'))
+        self.extract(use_hpss, zstash_dir)
+        msg = 'Deleting the extracted files and doing it again without verbose option, '
+        msg += 'while making sure the tars are kept.'
+        print_starred(msg)
+        self.assertWorkspace()
+        shutil.rmtree(self.test_dir)
+        os.mkdir(self.test_dir)
+        os.chdir(self.test_dir)
+        if not use_hpss:
+            shutil.copytree('{}/{}/zstash'.format(TOP_LEVEL, self.backup_dir), self.copy_dir)
+        keep_option = ' --keep' if (keep == 'extract') else ''
+        cmd = '{}zstash extract --hpss={}{}'.format(zstash_dir, self.hpss_path, keep_option)
+        output, err = run_cmd(cmd)
+        # Check that the zstash/ directory contains all expected files
+        if not compare(os.listdir(self.copy_dir),
+                       ['index.db', '000000.tar', '000001.tar', '000002.tar', '000003.tar', '000004.tar']):
+            error_message = 'The zstash directory does not contain expected files.\nIt has: {}'.format(
+                os.listdir(self.copy_dir))
+            self.stop(error_message)
+        os.chdir(TOP_LEVEL)
+        expected_present = ['Extracting file0.txt', 'Extracting file0_hard.txt', 'Extracting file0_soft.txt',
+                            'Extracting file_empty.txt', 'Extracting dir/file1.txt', 'Extracting empty_dir',
+                            'Extracting dir2/file2.txt', 'Extracting file3.txt', 'Extracting file4.txt',
+                            'Extracting file5.txt']
+        if use_hpss:
+            expected_present.append('Transferring file from HPSS')
+        expected_absent = ['ERROR', 'Not extracting']
+        self.check_strings(cmd, output + err, expected_present, expected_absent)
+
+    def testExtractVerbose(self):
+        self.helperExtractVerbose('testExtractVerbose', 'none')
+
+    @unittest.skipIf(os.system('which hsi') != 0, 'This system does not have hsi')
+    def testExtractVerboseHPSS(self):
+        self.helperExtractVerbose('testExtractVerboseHPSS', 'zstash_test')
+
+    def testKeepOnExtract(self):
+        self.helperKeep('testKeepOnExtract', 'none', keep='extract')
+
+    @unittest.skipIf(os.system('which hsi') != 0, 'This system does not have hsi')
+    def testKeepOnExtractHPSS(self):
+        self.helperKeep('testKeepOnExtractHPSS', 'zstash_test', keep='extract')
+
+    def testKeepOnCreate(self):
+        self.helperKeep('testKeepOnCreate', 'none', keep='create')
+
+    #@unittest.skipIf(os.system('which hsi') != 0, 'This system does not have hsi')
+    #def testKeepOnCreateHPSS(self):
+    #    self.helperKeep('testKeepOnCreateHPSS', 'zstash_test', keep='create')
```

Comment:
@golaz This tests fails. I assume this to be because the test is bad, not because the actual code is bad. The test assumes behavior is the same as `zstash extract --keep` except the `--keep` is instead applied to `zstash create`.